src/sat.cr

```diff
@@ -0,0 +1,207 @@
+# Crystal port of simple-sat by Sahand Saba distributed under the MIT License.
+# - https://github.com/sahands/simple-sat/
+# - https://sahandsaba.com/understanding-sat-by-implementing-a-simple-sat-solver-in-python.html
+
+module Shards
+  class SAT
+    alias Literal = Int32
+    alias Clause = Array(Literal)
+
+    def self.from_io(io : IO) : SAT
+      new.tap do |sat|
+        while line = io.gets
+          next if line.empty? || line.starts_with?('#')
+          sat.add_clause(line)
+        end
+      end
+    end
+
+    def self.from_file(path : String) : SAT
+      File.open(path) { |f| from_io(f) }
+    end
+
+    def initialize
+      @clauses = Array(Clause).new
+      @table = Hash(String, Literal).new
+      @variables = Array(String).new
+    end
+
+    def add_clause(str : String) : Nil
+      add_clause(str.split)
+    end
+
+    def add_clause(ary : Array(String)) : Nil
+      clause = ary.map do |literal|
+        negated = literal.starts_with?('~') ? 1 : 0
+        variable = literal[negated..-1]
+
+        unless @table.has_key?(variable)
+          @table[variable] = @variables.size
+          @variables << variable
+        end
+
+        @table[variable] << 1 | negated
+      end
+
+      clause.uniq!
+      @clauses << clause
+    end
+
+    protected def literal_to_s(literal : Literal) : String
+      String.build { |str| literal_to_s(str, literal) }
+    end
+
+    protected def literal_to_s(io : IO, literal : Literal) : Nil
+      io << '~' unless literal & 1 == 0
+      io << @variables[literal >> 1]
+    end
+
+    protected def clause_to_s(clause : Clause) : String
+      String.build do |str|
+        clause.each_with_index do |literal, index|
+          str << ' ' unless index == 0
+          literal_to_s(str, literal)
+        end
+      end
+    end
+
+    protected def assignment_to_s(assignment : Array(Literal?), brief = false)
+      String.build do |str|
+        assignment.zip(@variables).compact_map do |(a, v)|
+          if a == 0
+            next if brief
+            str << '~'
+          end
+          str << v << ' '
+        end
+      end
+    end
+
+    protected def to_assigment(assignment : Array(Literal?), brief = false)
```

Comment:
This method is super slow and must be optimized. The solve method below can find 1 million solutions out of 300 variables and 5000 clauses in 0.8s (release mode) which is acceptable performance; calling this method to transform the assignment array to an array of variables easily takes more than 10s :fearful: